src/Npgsql/NpgsqlConnection.cs

```diff
@@ -1203,22 +1203,18 @@ public bool Wait(int timeout)
         /// (see https://stackoverflow.com/questions/12421989/networkstream-readasync-with-a-cancellation-token-never-cancels ).
         /// </summary>
         [PublicAPI]
-        public Task WaitAsync(CancellationToken cancellationToken)
+        public Task WaitAsync(CancellationToken cancellationToken = default)
         {
+            if (cancellationToken.IsCancellationRequested)
+                return Task.FromCanceled(cancellationToken);
```

Comment:
There is a tiny semantical difference here though. 

`Task.FromCanceled` will create a `Task.Status == Canceled` with `TaskCanceledException`.
While throwing from within an ordinary async method would transition the task to Faulted with an `OperationCanceledException` (which is the base class for `TaskCanceledException`) because the task wasn't started with the passed in token.

Reply:
I may be misunderstanding, but consider the following:

```c#
[Test]
public void ThrowInAsyncAfterYield()
{
    var t = Foo(new CancellationToken(canceled: true));
    Assert.That(async () => await t, Throws.Exception.TypeOf<TaskCanceledException>());
    Assert.That(t.Status, Is.EqualTo(TaskStatus.Canceled));

    static async Task Foo(CancellationToken token)
    {
        await Task.Yield();
        token.ThrowIfCancellationRequested();
    }
}

[Test]
public void ReturnTaskCanceledInNonAsync()
{
    var t = Foo(new CancellationToken(canceled: true));
    Assert.That(async () => await t, Throws.Exception.TypeOf<TaskCanceledException>());
    Assert.That(t.Status, Is.EqualTo(TaskStatus.Canceled));

    static Task Foo(CancellationToken token)
        => token.IsCancellationRequested
            ? Task.FromCanceled(token)
            : Task.CompletedTask;
}
```

ThrowInAsyncAfterYield's task is ultimately Canceled, not Faulted.

You may also want to look at https://github.com/dotnet/corefx/pull/38453 and the conversation on it.

Reply:
@NinoFloris anything more to say here?

Reply:
No sorry, have been meaning to reply. But it's interesting it works this way. Does that TaskCanceled exception also include that cancellationtoken from the arguments? In that case I'm really curious what magic the compiler does there.

Reply:
You're asking where (or whether) the token is stored on cancelled tasks, right? It could internally be a union with the exception that needs to be recorded on faulted tasks, for example...

Otherwise `token.ThrowIfCancellationRequested()` simply goes `throw new OperationCanceledException(token)`, which gets caught and results in the task being Cancelled, as far as I understand it.